pyccel/codegen/printing/fcode.py

```diff
@@ -2608,31 +2608,77 @@ def _print_IndexedElement(self, expr):
+            _shape = PyccelArraySize(base, i if expr.order != 'C' else len(inds) - i - 1)
+            if isinstance(ind, Slice):
+                inds[i] = self._slice_management(ind, _shape, allow_negative_indexes)
+            elif isinstance(ind, PyccelUnarySub) and isinstance(ind.args[0], LiteralInteger):
+                inds[i] = PyccelMinus(_shape, ind.args[0])
             else:
                 #indices of indexedElement of len==1 shouldn't be a Tuple
                 if isinstance(ind, Tuple) and len(ind) == 1:
                     inds[i] = ind[0]
                 if allow_negative_indexes and not isinstance(ind, LiteralInteger):
-                    inds[i] = PyccelMod(ind, base_shape[i])
+                    inds[i] = PyccelMod(ind, _shape)
 
         inds = [self._print(i) for i in inds]
 
-        return "%s(%s)" % (base, ", ".join(inds))
+        return "%s(%s)" % (base_code, ", ".join(inds))
 
 
     def _print_Idx(self, expr):
         return self._print(expr.label)
 
+    def _slice_management(self, _slice, shape, allow_negative_index):
+        start = _slice.start
+        end = _slice.end
+        step = _slice.step
+
+        # negative start and end in slice
+        if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
+            start = PyccelMinus(shape, start.args[0])
+        elif start is not None and allow_negative_index and not isinstance(start,LiteralInteger):
+            start = IfTernaryOperator(PyccelLt(start, LiteralInteger(0)),
+                        PyccelAdd(shape, start), start)
+
+        if isinstance(end, PyccelUnarySub) and isinstance(end.args[0], LiteralInteger):
+            end = PyccelMinus(shape, end.args[0])
+        elif end is not None and allow_negative_index and not isinstance(end, LiteralInteger):
+            end = IfTernaryOperator(PyccelLt(end, LiteralInteger(0)),
+                        PyccelAdd(shape, end), end)
+
+        tmp_end = end #save temporary value of end
+        # steps in slices
+        if step is not None :
+            #negative step in slice
+            if isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
+                end = PyccelAdd(end, LiteralInteger(1)) if end is not None else LiteralInteger(0)
+                start = start if start is not None else PyccelMinus(shape, LiteralInteger(1))
+
+            # variable step in slice
+            elif allow_negative_index and not isinstance(step, LiteralInteger):
+                if start is None :
+                    start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)),
+                        LiteralInteger(0), PyccelMinus(shape, LiteralInteger(1)))
+
+                if end is None :
+                    end = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)),
+                        PyccelMinus(shape, LiteralInteger(1)), LiteralInteger(0))
+                else :
+                    end = PyccelAdd(end, LiteralInteger(1))
```

Comment:
I am confused. We have:
```
0:10
```
translated to:
```
0:11
```
but 
```
0:
```
translated to:
```
0:shape-1
```
?

Reply:
this is when giving 
```
a = b[ : : step]
```
- first point .
here Fortran manage step differently than python when end and start are none
writing :  
```
a = b[ : : step] # if step > 0 this will work
a = b[ : : step] # if step < 0 this will not  work so we need to prisise the start and the end for fortran
```
- second point.
if we take a look at python slice it looks like this [ start: end [  end is excluded but in Fortran [start, end] end is included so when we have  
```
a = b [ : 2: -2] to exulde 2 I add 1.
```

Reply:
I am happy with the first case, I am just confused about the second case.

Here we are handling a case such as `b[0:2:i]`. If `i=1` then python will create an array containing `b[0]`, and `b[1]` but with the same syntax fortran will create an array containing `b[0]`, `b[1]` and `b[2]` so we need to print `b[0:1:i]` to get the same result.

As far as I can tell this case would fall to line 2675 so we would print `b[0:3:i]`. Should this not be a ternary operator?

I think this makes sense for `b[4:2:i]` with `i=-1`,  then python will create an array containing `b[3]`, and `b[4]` but with the same syntax fortran will create an array containing `b[2]`, `b[3]` and `b[4]` so we need to print `b[4:3:i]` to get the same result.

Reply:
seem's logic. I'm gonna remove it once I remember why I put it in the first place.